apex/amp/utils.py

```diff
@@ -92,9 +92,12 @@ def cached_cast(cast_fn, x, cache):
         return type(x)([cached_cast(y) for y in x])
     if x in cache:
         cached_x = cache[x]
+        next_functions_available = False
         if x.requires_grad and cached_x.requires_grad:
+            if len(cached_x.grad_fn.next_functions) > 1:
+                next_functions_available = True
             # Make sure x is actually cached_x's autograd parent.
-            if cached_x.grad_fn.next_functions[1][0].variable is not x:
+            if next_functions_available and cached_x.grad_fn.next_functions[1][0].variable is not x:
```

Comment:
```suggestion
            if len(cached_x.grad_fn.next_functions) > 1 and cached_x.grad_fn.next_functions[1][0].variable is not x:
```
Would this be less change? And maybe a comment explain why we need this

Reply:
Nvm, I didn't see `next_functions_available` is used again below. Maybe just a comment to explain why we need this change

Reply:
This change is not enough, because we need ensure it `del cache[x]` in this case.